### PR TITLE
docker: implement new design

### PR DESCRIPTION
**Pull request by larskarlitski**
Working towards the new [design][1] for the docker page.

- [x] row headers in cockpit's current listing view react component look different from the design
- [x] hook up buttons in expanded container row
- ~~[ ] add console and metrics tabs to expanded container row~~
- ~~[ ] maybe add "containers" tab to expanded image row~~
- [x] figure out where to put "get new image" button
- [x] add filtering by name
- ~~[ ] change storage and metric cards~~
- [x] adjust tests

(Edit: postpone some of the goals here to get the bulk of this rewrite in and unblock things.)

screenshot from 2016-09-01 17-54-57

[1]: https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/containers/container-list-design.png

### CODE REVIEW

**Comment by larsu** [comment]
Depends on #5043.

**Comment by larskarlitski** [comment]
Update this:

- remove the listingview commits that @dperpeet merged
- rename some ids to make tests work (still some issues left I think, so not untagging `needswork`)
- use @dperpeet's select component instead of a custom one (makes this branch depend on #5156)
- add filter box
- fix a few minor bugs

**Comment by larskarlitski** [comment]
Update: fix tests.

**Comment by larskarlitski** [comment]
The remaining test failures look unrelated to me:

- most of them are in different modules entirely
- the rhel ones fail on a `{ transition }` selinux denial
- I cannot debug the testRestart one in fedora-25, because it doesn't have a log (and I can't reproduce it on my machine)

File: pkg/docker/containers-view.jsx [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+    stopContainer: function (container) {
+        this.props.client.stop(container.Id).fail(util.show_unexpected_error);
+    },
+
+    restartContainer: function (container) {
+        this.props.client.restart(container.Id).fail(util.show_unexpected_error);
+    },
+
+    deleteContainer: function (container) {
+        if (event.button !== 0)
```
Comment (stefwalter) [inline pkg/docker/containers-view.jsx]: event is not defined.

Comment (larskarlitski) [inline pkg/docker/containers-view.jsx]: Weird, WebKit has a global `event`, which makes this work. Thanks for the catch. Fixed.

**Comment by larskarlitski** [comment]
This is still blocked on #5043.

**Comment by larskarlitski** [comment]
Rebased on master, which contains #5156 now.

File: pkg/docker/containers-view.jsx [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    }
+});
+
+var ContainerDetails = React.createClass({
+    render: function () {
+        var container = this.props.container;
+        return (
+            <div className='listing-body'>
+                <dl>
+                    <dt>Id       </dt> <dd>{ container.Id }</dd>
```
Comment (stefwalter) [inline pkg/docker/containers-view.jsx]: These strings need to be translated.

File: pkg/docker/containers-view.jsx [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+                                </a>;
+
+        var columnTitles = [ _('Name'), _('Created'), _('Size'), '' ];
+
+        var pendingRows = this.state.pulling.map(function (job) {
+            if (job.error)
+                return <p className="status has-error">{job.error}</p>;
+
+            var detail = '';
+            if (job.progress)
```
Comment (stefwalter) [inline pkg/docker/containers-view.jsx]: This needs braces to be more clear.

File: pkg/docker/containers.js [hunk 4]
```diff
@@ -139,6 +139,7 @@
         cockpit.translate();
 
         docker_client = client.instance();
+
```
Comment (stefwalter) [inline pkg/docker/containers.js]: Unnecessary change?

File: pkg/docker/search.js [hunk 5]
```diff
@@ -86,6 +90,9 @@
             var registry = $('#containers-search-download').data('registry') || undefined;
             var tag = $('#containers-search-tag').val();
 
+            this.dfd.resolve(repo, tag, registry);
+
+            /*
```
Comment (stefwalter) [inline pkg/docker/search.js]: Why is this commented out? Maybe just remove it?

**Comment by larskarlitski** [comment]
Update: fix issues that @stefwalter raised. Thanks!

File: pkg/docker/search.js (hunk 5 above)
Comment (larskarlitski) [inline pkg/docker/search.js]: Ugh, that one's embarrassing. Thanks :)

**Comment by dperpeet** [comment]
The filter input box is misaligned: 
screenshot from 2016-10-18 22-49-31

**Comment by dperpeet** [comment]
I think the filtering should also apply to the images, not just the running containers. Otherwise the layout isn't clear to me.

**Comment by dperpeet** [comment]
There should be some indication how many items were filtered, this also includes non-running containers that are hidden by default. Maybe a button like in GitHub to clear filters and show everything?

File: pkg/docker/containers-view.jsx [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+        var action = this.props.actions[event.currentTarget.getAttribute('data-value')];
+        if (!action.disabled && action.onActivate)
+            action.onActivate();
+    },
+
+    render: function () {
+        return (
+            <div className="btn-group">
+                <button className="btn btn-default" type="button" data-value="0" onClick={this.handleClick}>
+                    <span>{this.props.actions[0].label}</span>
```
Comment (dperpeet) [inline pkg/docker/containers-view.jsx]: didn't we want to add spaces to non-trivial content of `{ }`?

File: pkg/docker/containers-view.jsx [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+                </button>
+                <button className="btn btn-default dropdown-toggle" data-toggle="dropdown">
+                    <div className="caret"></div>
+                </button>
+                <ul className="dropdown-menu dropdown-menu-right" role="menu">
+                    {
+                        this.props.actions.map(function (action, index) {
+                            return (
+                                <li className={action.disabled ? 'disabled' : ''}>
```
Comment (dperpeet) [inline pkg/docker/containers-view.jsx]: didn't we want to add spaces to non-trivial content of { }?

File: pkg/docker/containers-view.jsx [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+                                       navigateToItem={this.navigateToContainer.bind(this, container)}
+                                       listingActions={actions}/>;
+        }, this);
+
+        var columnTitles =  [ _('Name'), _('Image'), _('Command'), _('CPU'), _('Memory'), _('State')];
+
+        return (
+            <Listing.Listing title={_('Containers')}
+                             columnTitles={columnTitles}
+                             emptyCaption={this.props.onlyShowRunning ? _("No running containers") : _("No containers")}>
```
Comment (dperpeet) [inline pkg/docker/containers-view.jsx]: didn't we want to add spaces to non-trivial content of { }?

File: pkg/docker/containers-view.jsx [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+
+    render: function () {
+        var imageRows = this.state.images.map(function (image) {
+            var columns = [
+                { name: image.RepoTags[0], header: true },
+                moment.unix(image.Created).fromNow(),
+                cockpit.format_bytes(image.VirtualSize),
+                {
+                    element: <button className="btn btn-default btn-control-ct fa fa-play"
+                                     onClick={this.showRunImageDialog.bind(this)}
```
Comment (dperpeet) [inline pkg/docker/containers-view.jsx]: didn't we want to add spaces to non-trivial content of { }?

File: pkg/docker/containers-view.jsx [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+                {
+                    name: _('Details'),
+                    renderer: ImageDetails,
+                    data: { image: image }
+                }
+            ];
+
+            var actions = [
+                <button className="btn btn-danger btn-delete pficon pficon-delete"
+                        onClick={this.deleteImage.bind(this, image)} />
```
Comment (dperpeet) [inline pkg/docker/containers-view.jsx]: didn't we want to add spaces to non-trivial content of { }?

File: pkg/docker/containers-view.jsx [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+            ];
+
+            return <Listing.ListingRow key={image.Id}
+                                       columns={columns}
+                                       tabRenderers={tabs}
+                                       navigateToItem={this.navigateToImage.bind(this, image)}
```
Comment (dperpeet) [inline pkg/docker/containers-view.jsx]: didn't we want to add spaces to non-trivial content of { }?

**Comment by dperpeet** [comment]
Also, but this probably needs to go into a follow-up: it is extremely cumbersome right now to delete multiple containers. This is made more difficult by the fact that you need to click the expand arrow exactly, otherwise you navigate to another page, and after expanding the delete button is on the other side of the screen.

**Comment by andreasn** [comment]
> Also, but this probably needs to go into a follow-up: it is extremely cumbersome right now to delete multiple containers.

It's an obvious regression, but maybe that's OK for now.
See screen 3 for how this could work.
https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/containers/container-list-design.png

**Comment by andreasn** [comment]
> There should be some indication how many items were filtered, this also includes non-running containers that are hidden by default. Maybe a button like in GitHub to clear filters and show everything?

Do you mean something like this?
screenshot from 2016-10-19 11-29-32

I think that could work.

**Comment by andreasn** [comment]
I seem to get a lot of these Oops:es

screenshot from 2016-10-19 11-33-31

**Comment by larskarlitski** [comment]
Pushed a fix for the style isseus @dperpeet mentioned. Thanks!

> I think the filtering should also apply to the images, not just the running containers. Otherwise the layout isn't clear to me.

I agree that having the filter controls on top of the page is misleading, because they only apply to containers. Even if images were filtered as well, the resource usage graphs would still be between the controls and the results. @andreasn, what do you think?

> There should be some indication how many items were filtered, this also includes non-running containers that are hidden by default. Maybe a button like in GitHub to clear filters and show everything?

We don't have as many filter settings as github does. I think the real issue here is that the filter settings are too far away from the elements they're filtering.

That said, I think changing the empty state of the list views to "No [running] containers that match the current filter" would go a long way to avoid confusion. What do you think?

**Comment by andreasn** [comment]
> That said, I think changing the empty state of the list views to "No [running] containers that match the current filter" would go a long way to avoid confusion. What do you think?

I like this soultion

**Comment by andreasn** [comment]
> I agree that having the filter controls on top of the page is misleading, because they only apply to containers.

I think this could be reinforced with wording. What about:
 * All -> Images and containers
 * Running -> Images and running containers only

**Comment by andreasn** [comment]
Taking a step back, I think hiding the graphs could be a possibility in the filtering, but it would at the same time go against these use cases:
 1. I'm running out of memory/CPU due to too many containers running. I need to delete or stop some containers. Seeing the graphs helps me get an idea about the total amount of memory/CPU I'm using.
 2. I'm running out of space. I need to delete some images. (maybe some container that aren't running too?). Seeing the graphs helps me get an idea about the total amount of storage used.
 3. Is there a third filter case I'm forgetting about?

Putting the filters elsewhere on the page could be another possibility, but it's very convenient from a layout perspective to have it at the top. Especially if we reinforce this filter pattern elsewhere.

**Comment by andreasn** [comment]
Another possible followup. Putting the graphs and storage in Cards.
http://www.patternfly.org/pattern-library/cards/aggregate-status-card/#/_code

**Comment by andreasn** [comment]
> The filter input box is misaligned:

It appears the dropdown and input have different ideas how far from the top it's appropriate for a control to be.
screenshot from 2016-10-19 12-34-22

**Comment by larskarlitski** [comment]
> I like this soultion

Thanks. Done.

> I think this could be reinforced with wording. What about:

Just tried those. They make the combo box uncomfortably wide. How about "All containers" and "Only running containers" instead?

> Taking a step back, I think hiding the graphs could be a possibility in the filtering, but it would at the same time go against these use cases: [...]

I agree that these use cases are important. Also, I think removing elements from the page when searching might just add to the confusion.

> Putting the filters elsewhere on the page could be another possibility, but it's very convenient from a layout perspective to have it at the top. Especially if we reinforce this filter pattern elsewhere.

Good point about consistency. Maybe this problem will be alleviated a bit when we have the cards?

**Comment by andreasn** [comment]
> Just tried those. They make the combo box uncomfortably wide. How about "All containers" and "Only running containers" instead?

Ugh. Problem is that it also shows images. On the other hand, German. Another solution
 * Everything
 * Images & running containers

The second is 4 letters longer than "Only running containers" It's a bit backwards that images are before containers, even though they come after on the page, but otherwise the sentence starts referring to running images.

**Comment by larskarlitski** [comment]
>  Another solution [...]

Cool, I like that one. Done.

> On the other hand, German.

Haha!

**Comment by larskarlitski** [comment]
> I seem to get a lot of these Oops:es

Thanks for the catch. This happens when hovering over the CPU or memory graphs, because it would highlight the corresponding row in the table. I've tried to replicate that behavior, but the current listing view doesn't allow highlighting rows from the outside. Thus, I've removed this feature for now, but I think we should revisit it in the future.

**Comment by stefwalter** [comment]
Here's a commit that you need in order to make rhel-7 pass: https://github.com/stefwalter/cockpit/commit/01b7bd8cd36590c6630d5b8299b857901cdf152c

**Comment by larskarlitski** [comment]
> Here's a commit that you need in order to make rhel-7 pass: stefwalter@01b7bd8

Thanks! Pushed on here.

**Comment by dperpeet** [comment]
Looks good to me. If tests (especially rhel-7) pass, I think this is good to go.

**Comment by dperpeet** [comment]
I just tried deleting images and got errors
```
Object { status: 500, reason: "Internal Server Error", message: "layer does not exist ", problem: null, valueOf: HttpError/this.valueOf(), toString: HttpError/this.toString() }
```
Was this a local fluke?

**Comment by larskarlitski** [comment]
> Then the docker page gets disconnected.

I've tried a a number of different ways but cannot reproduce this. Which version of docker are you using?

**Comment by stefwalter** [comment]
> Then the docker page gets disconnected.

That's a separate bug. It happened with the old page. Docker returns 500 errors, which we in turn disconnect on. This should be a follow up.

**Comment by stefwalter** [comment]
Selenium tests look like they need an update.

**Comment by dperpeet** [comment]
> Selenium tests look like they need an update.

Already looking

**Comment by dperpeet** [comment]
On my test machine, the list of images isn't refreshed after deleting an image or downloading a new one.

**Comment by dperpeet** [comment]
This commit fixes the selenium docker test: https://github.com/dperpeet/cockpit/commit/f3e5a771f945da25a4d53fb249bb7466eacf059f
